File: firebase/functions/lib/events/notifications/send_event_message.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart'
    as functions_interop;
import '../../on_call_function.dart';
import '../../utils/email_templates.dart';
import '../../utils/infra/firestore_utils.dart';
import '../../utils/notifications_utils.dart';
import '../../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';


import '../../utils/template_utils.dart';

class SendEventMessage extends OnCallMethod<SendEventMessageRequest> {
  static const String kSendEventMessageApi = 'sendEventMessage';
  NotificationsUtils notificationsUtils;

  SendEventMessage({NotificationsUtils? notificationsUtils})
      : notificationsUtils = notificationsUtils ?? NotificationsUtils(),
        super(
          kSendEventMessageApi,
          (jsonMap) => SendEventMessageRequest.fromJson(jsonMap),
        );

  @override
  Future<void> action(
    SendEventMessageRequest request,
    functions_interop.CallableContext context,
  ) async {
    orElseUnauthorized(context.authUid != null);

    final documentData = DocumentData.fromMap(
      firestoreUtils.toFirestoreJson(request.eventMessage.toJson()),
    );

    final createdAtMillis = request.eventMessage.createdAtMillis ??
        DateTime.now().millisecondsSinceEpoch;
    final createdAtTimestamp = Timestamp.fromDateTime(
      DateTime.fromMillisecondsSinceEpoch(createdAtMillis),
    );

    documentData.setTimestamp('createdAt', createdAtTimestamp);

    print('Adding data: ${documentData.toMap()}');

    // First of all - add a new event message to the firestore
    await firestore
        .collection('community')
        .document(request.communityId)
        .collection('templates')
        .document(request.templateId)
        .collection('events')
        .document(request.eventId)
        .collection('event-messages')
        .add(documentData);

    // Then get template and event in one go, to make it faster
    final eventDocFuture = firestore
        .collection('community')
        .document(request.communityId)
        .collection('templates')
        .document(request.templateId)
        .collection('events')
        .document(request.eventId)
        .get();
    final templateDocFuture = firestore
        .collection('community')
        .document(request.communityId)
        .collection('templates')
        .document(request.templateId)
        .get();

    final List<DocumentSnapshot> futures = await Future.wait<DocumentSnapshot>([
      templateDocFuture,
      eventDocFuture,
    ]);

    // Parse it after both futures completed
    final template = TemplateUtils.templateFromSnapshot(futures[0]);
    final event = Event.fromJson(
      firestoreUtils.fromFirestoreJson(futures[1].data.toMap()),
    );

    await notificationsUtils.sendEmailToEventParticipants(
      communityId: request.communityId,
      template: template,
      event: event,
      generateMessage: ({
        required community,
        required user,
        required unsubscribeUrl,
      }) {
        return SendGridEmailMessage(
          subject: 'New Message in Event ${event.title}',
          html: makeNewEventMessageBody(
            community: community,
            template: template,
            event: event,
            eventMessage: request.eventMessage,
            unsubscribeUrl: unsubscribeUrl,
          ),
        );
      },
    );
  }
}

File: firebase/functions/src/events/notifications/send_event_message.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { notificationsUtils } from '../../utils/notifications_utils';
import { makeNewEventMessageBody } from '../../utils/email_templates';
import { orElseUnauthorized } from '../../utils/utils';
import { TemplateUtils } from '../../utils/template_utils';
import { Event, EventMessage } from '../../types';

interface SendEventMessageRequest {
  communityId: string;
  templateId: string;
  eventId: string;
  eventMessage: EventMessage;
}

export class SendEventMessage extends OnCallMethod<SendEventMessageRequest> {
  constructor() {
    super('sendEventMessage', (jsonMap) => jsonMap as unknown as SendEventMessageRequest);
  }

  async action(request: SendEventMessageRequest, context: functions.https.CallableContext): Promise<void> {
    orElseUnauthorized(context.auth?.uid != null);

    const msgData: Record<string, unknown> = {
      ...(firestoreUtils.toFirestoreJson(request.eventMessage as unknown as Record<string, unknown>)),
      createdAt: new Date(request.eventMessage.createdAtMillis ?? Date.now()),
    };

    await firestore
      .collection(`community/${request.communityId}/templates/${request.templateId}/events/${request.eventId}/event-messages`)
      .add(msgData);

    const [templateDoc, eventDoc] = await Promise.all([
      firestore.doc(`community/${request.communityId}/templates/${request.templateId}`).get(),
      firestore.doc(`community/${request.communityId}/templates/${request.templateId}/events/${request.eventId}`).get(),
    ]);

    const template = TemplateUtils.templateFromSnapshot(templateDoc);
    const event: Event = {
      ...(firestoreUtils.fromFirestoreJson(eventDoc.data() ?? {}) as unknown as Event),
      id: eventDoc.id,
    };

    await notificationsUtils.sendEmailToEventParticipants({
      communityId: request.communityId,
      template,
      event,
      generateMessage: ({ community, user, unsubscribeUrl }) => ({
        subject: `New Message in Event ${event.title}`,
        html: makeNewEventMessageBody({
          community,
          template,
          event,
          eventMessage: request.eventMessage,
          unsubscribeUrl,
        }),
      }),
    });
  }
}
